wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -245,6 +246,72 @@ public void stop(String workspaceId) throws NotFoundException, ServerException,
     }
 
     /**
+     * Adds machine into running workspace.
+     * This method do not touch workspace configuration.
+     * Just adds machine to workspace runtime and destroy it on workspace stop.
+     * Does nothing if add already existing machine.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     * @throws ServerException
+     *         when application server is stopping
+     * @throws ConflictException
+     *         when workspace is not running
+     */
+    public void addMachine(MachineImpl machine) throws NotFoundException, ServerException, ConflictException {
```

Comment:
Implementation class is the right choice here.
Explanation:
Lets start from the [Machine](https://github.com/eclipse/che/blob/master/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Machine.java) interface.

```java
public interface Machine {
   // ...
   MachineConfig getConfig();
   // ...
}
```

This interface defines `MachineConfig`(which is also model interface) as configuration type.
So we have several outcomes:
-  Dto can't use non-dto classes, which means that MachineDto implementation must return `MachineConfigDto` instead, which means that all the classes used in MachineDto are dtos. So we override the method and define 
[MachineConfigDto](https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineDto.java#L28) as return type of the `config()` method
- `MachineImpl` follows the same strategy and uses Impl objects, instead of model objects in composition.
     - It allows to use setters(not defined by model), which is really important
     - There is no any reason to define the composition of the object from the same family using interfaces, because
it is not profitable at all, it complicates process of changing values and doesn't give the benefits from the abstraction, 
and it is supposed that objects family is used as a one integrated piece.

But why don't we use the interfaces for such classes like managers or daos?
Explanation:
1. We don't have data object factories or abstract builders, which means that we create objects directly by using `new`. In this case managers and daos are automatically dependent from the `*Impl` classes.
2. We need to modify(use setters) data returned by daos and managers, so in the case of model interfaces we need to recreate object over and over again adding new modifications to it, but what is the profit?

So we decided to use *Impl* objects throughout the same workspaces infrastructure, but use model objects in dependent parts e.g. `Account`. It allows us to avoid allocations and clears relationship between workspace internals. 

For DTOs to Impl objects transformation we define copy constructors in `Impl` classes based on model interfaces, which allows to switch between implementations easily.

In this case Model object is rather for integrity and consistency of client and server, than for something else.